src/routing/kad-dht.md

```diff
@@ -0,0 +1,719 @@
+---
+title: IPFS Kademlia DHT
+description: >
+  The IPFS Distributed Hash Table (DHT) specification defines a structured
+  overlay network used for peer and content routing in the InterPlanetary File
+  System (IPFS). It extends the libp2p Kademlia DHT specification, adapting and
+  adding features to support IPFS-specific requirements.
+date: 2025-03-24
+maturity: reliable
+editors:
+  - name: Guillaume Michel
+    url: https://guillaume.michel.id
+    github: guillaumemichel
+    affiliation:
+      name: Shipyard
+      url: https://ipshipyard.com
+tags: ['routing']
+order: 1
+---
+
+The IPFS Distributed Hash Table (DHT) specification defines a structured
+overlay network used for peer and content routing in the InterPlanetary File
+System (IPFS). It extends the libp2p Kademlia DHT specification, adapting and
+adding features to support IPFS-specific requirements.
+
+## Introduction
+
+The Kademlia Distributed Hash Table (DHT) is a decentralized key-value store
+designed to enable efficient and scalable peer-to-peer routing. It provides a
+structured overlay network that allows nodes to locate peers and content in a
+distributed system without relying on centralized servers.
+
+The primary goal of the Kademlia routing algorithm is to progressively discover
+and interact with nodes that are closest to a given key based on the network's
+distance metric. Once a node has identified the closest peers, it can either:
+
+* **Locate a specific peer** in the network
+* **Find content providers** serving content associated with a CID
+* **Store and retrieve values** directly within the DHT, such as IPNS names
+
+### Relation to libp2p kad-dht
+
+The IPFS Kademlia DHT specification is a specialization of the [libp2p Kademlia DHT](https://github.com/libp2p/specs/tree/master/kad-dht).
+
+It is possible to use an alternative DHT specification alongside an IPFS
+implementation, rather than the one detailed here. This document specifically
+outlines all protocol customizations and adaptations required for participation
+in the [Amino DHT](#relation-to-the-amino-dht). If you're designing a new
+Kademlia-based DHT for use with IPFS, some details in this specification may
+appear overly specific or prescriptive.
+
+### Relation to the Amino DHT
+
+Nodes participating in the public [Amino DHT Swarm](#amino-dht) MUST implement the
```

Comment:
Thoughts on how we should handle parts of the network that evolve over time?

For example:
- Maybe we mandate TCP + Yamux + Noise for all DHT server nodes, or we mandate TCP + Yamux + Noise and QUIC for all DHT client nodes and say servers only need to support one. We could then encourage with SHOULDs other transports like WSS, WT, WebRTC, etc. and change from SHOULD to MUST as we see adoption rise past a certain point where we'd feel comfortable having clients rely on the new behavior from the servers.
- Same if we decide to add new RPC methods / behaviors or change some server behavior without bumping the protocol version (e.g. IP / ASN limiting, etc.)

Reply:
Good points!

* I will add transports requirements
* For new RPC methods/behaviors, I suggest we deal with it in a subsequent PR/IPIP since it will require some discussion